### PR TITLE
Redesign Entity and Algorithm classes

### LINKED ISSUE #58: Entity requires additional initialization apart from ctor invocation
`Entity` class has method `calculateVector` which initializes field `vector` and (as far as I can see) is intended to be called as part of an initialization process.

### LINKED ISSUE #65: Entity abstract class heavily depends on its subclasses
`Entity` class knows about all its subclasses and stores `VectorCalculator` for each of them. It is done to fulfil requirement that features vectors for entities should have equal dimension.

### PR DESCRIPTION
**Pull request by Ivan-Veselov**
Changes current design of `Entity` and `Algorithm` classes.
`Entity` is replaced by `CodeEntity` and `ElementAttributes` and now its responsibility is spreaded across several classes.
Single `Algorithm` class is replaced by chain of `Algorithm` interface and abstract `AbstractAlgorithm` class. Now algorithm itself says calculation of what metrics it requires.

Old classes are left for backward compatibility with current implementation of algorithms but marked as deprecated.

Fixes #58 
Fixes #65

### CODE REVIEW

File: src/main/java/org/ml_methods_group/algorithm/attributes/AttributesStorage.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.List;
+
+/**
+ * One object of this class must be created per {@link Algorithm} execution. This class stores all
+ * relevant code entities (classes, methods, ...) along with their
+ * {@link org.ml_methods_group.algorithm.entity.RelevantProperties} and features extracted
+ * specially for one particular {@link Algorithm} from metrics this algorithm requested through
+ * {@link Algorithm#requiredMetrics}.
+ */
+public class AttributesStorage {
```
Comment (jzuken) [inline src/main/java/org/ml_methods_group/algorithm/attributes/AttributesStorage.java]: FeatureStorage?

File: src/main/java/org/ml_methods_group/algorithm/AbstractAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+     * services as parallel execution, progress report, etc.
+     */
+    protected final class ExecutionContext {
+        private final ExecutorService service;
+
+        private final ProgressIndicator indicator;
+
+        private final AttributesStorage attributes;
+
+        /**
```
Comment (jzuken) [inline src/main/java/org/ml_methods_group/algorithm/AbstractAlgorithm.java]: `/** ... */`

File: src/main/java/org/ml_methods_group/algorithm/attributes/ElementAttributes.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * new fresh {@link ElementAttributes}.
+ */
+public abstract class ElementAttributes {
+    /**
+     * Array of {@code double} values only to improve performance. It is not supposed to be changed.
+     */
+    private final @NotNull double[] features;
+
+    /**
+     * Initializes attributes.
```
Comment (jzuken) [inline src/main/java/org/ml_methods_group/algorithm/attributes/ElementAttributes.java]: `/** ... */`

File: src/main/java/org/ml_methods_group/algorithm/entity/ClassEntity.java [hunk 4]
```diff
@@ -4,46 +4,35 @@
...
-    private ClassEntity(ClassEntity original) {
-        super(original);
-        this.psiClass = original.psiClass;
-    }
-
     @Override
-    public MetricCategory getCategory() {
-        return MetricCategory.Class;
+    public @NotNull String getIdentifier() {
+        return psiClass.getQualifiedName();
```
Comment (jzuken) [inline src/main/java/org/ml_methods_group/algorithm/entity/ClassEntity.java]: а эта штука объявлена внутри как Nullable

File: src/main/java/org/ml_methods_group/algorithm/attributes/AttributesStorage.java (hunk 1 above)
Comment (Ivan-Veselov) [reply src/main/java/org/ml_methods_group/algorithm/attributes/AttributesStorage.java]: It is supposed that in the future `CodeEntity` and `RelevantProperties` will be separated and `RelevantProperties` will become more flexible like features currently are. With this separation `AttributesStorage` will have to store `RelevantProperties` as well (actually that was initial implementation). That is why it's called `AttributesStorage` where attributes refers to any kind of information that is derived from raw data.

File: src/main/java/org/ml_methods_group/algorithm/entity/ClassEntity.java (hunk 4 above)
Comment (Ivan-Veselov) [reply src/main/java/org/ml_methods_group/algorithm/entity/ClassEntity.java]: Ну если мы пользуемся знанием, что это класс, который прошел через `EntitySearcher`, то тут он должен быть `@NotNull`, иначе все сломается (много где используется идентификация по имени, там в базовом классе есть javadoc про это). В идеале этот метод вообще хочется убрать.

**Pull request merged**